joomla/joomla-cms
# [5.0] Fix custom field attribute 'showon' having no effect in the frontend

Pull Request for Issue #40762 and follow up of #39051.

### Summary of Changes

The use of the parameter 'showon' has been introduced in Joomla 4.3 for custom fields.
However, it lacks the handling of the fields when displayed in the front-end.
This PR is adding the missing functionality.

### Testing Instructions

In the 'Content' context:
Create a field of type 'text' named 'Conditional Text'.
Create a field of type 'list' named 'Box'. Add options (value1, value2 and value3).
Create a field of type 'list' named 'Square'. Add options (value1, value2 and value3).

For testing, you will need to change the value of the 'showon' attribute found in the 'options' tab for the text field.
After each change, open an article in the backend and change values of the Box and Square fields and check that the text field shows or hides depending on the values selected and the condition that has been set in the 'showon' attribute.
In the frontend, when opening the article, the display of the fields should match the display of the fields in the backend.

Values of the 'showon' attribute you can try out, as examples:
'Conditional Text' will show if:
`box:value1` => Box has value1
`box!:value1` => Box does not have value1
`box!:` => Box has a value that is not empty (in this case, you could add an empty option to that 'Box' field)
`box:value1[OR]square:value1` => Box has value1 or Square has value1
`box:value1[AND]square:value1` => Box and Square must have value1
`box:value1[OR]square!:value1` => Box has value1 or Square does not have value1
`box:value1[AND]square!:value1` => Box has value1 and Square has a value other than value1
`box:value1[AND]square:value1,value2` => Box has value1 and Square has value1 or value2
`box:value1[AND]square!:value1,value2` => Box has value1 and Square has value3

Test that wrong entries in the 'showon' parameter with values such as `tada`, `box::`, `box:!`, `box:3:21`, `[AND]box:value3[OR]tada:2:3`... does not generate errors in the front-end.

To test custom fields sent through emails, you will need to create fields in the 'Contact' - 'Mail ' context.
Create a field of type 'text' named 'Conditional Subject'.
Create a field of type 'list' named 'Sub'. Add options (value1, value2 and value3).
Make sure both fields have proper permissions to be editable in the contact form.
Set a 'showon' value for the text field, for example: `sub:value1`.
Now, when you go to the front-end and display the form for a contact, you should be able to switch between values of 'Sub' and check that the text field shows or not. When the email is sent, the content of that email should reflect the choices you made.

To test subforms, make 'Conditional Text', 'Box' and 'Square' part of a subform field.
Note that in subforms, the 'showon' parameter must use a different syntax, since the field names, in a subform, show as 'field[id]'. For instance, if 'Box' has id 15, the name of 'box', in the subform, will be 'field15'.
So you can try testing with 'showon' values such as  `field15:value1`.
Test the subform in the backend and the results in the frontend should match what has been selected in the backend.

Note: the 'showon' parameter will also 'work' with the syntax `{fieldgroup X}` that you can put inside the content of your article. However, it won't work if you hand-pick fields in the editor or have conditions set in different field groups. 

Notes to developers:

- why adding the code to match the showon parameter in FieldsHelper? So it can be made available globally and used in overrides and third-party extensions easily. The function names in the helper are not set in stone, I welcome suggestions.

- why have a particular case for subforms? Because when a subform is about to get displayed, the sub-fields have already been prepared for display and it's too late to change anything. Therefore, changes have to be made in the onCustomFieldsPrepareField function of the subform plugin.

- why make changes to the render.php files? So that we have all possible fields, prepared (it's also more efficient and prevents changes in too many places). The downside: we will only cover the `{fieldgroup X}` syntax of the content plugin, 'showon' won't work if fields are hand-picked or not in the same group.

### Actual result BEFORE applying this Pull Request

All fields show in the front-end, no matter the values that have been set in the 'showon' parameter.
Same goes for fields that will show in emails sent through the contact component.

### Expected result AFTER applying this Pull Request

Only the fields that are set to show actually show in the front-end.
Same goes for fields that will show in emails sent through the contact component.

### Link to documentations
For showon syntax, check: https://docs.joomla.org/Form_field#Showon
